domain/access/modelmigration/import.go

```diff
@@ -0,0 +1,109 @@
+
+package modelmigration
+
+import (
+	"context"
+	"time"
+
+	"github.com/juju/description/v8"
+	"github.com/juju/errors"
+
+	"github.com/juju/juju/core/logger"
+	coremodel "github.com/juju/juju/core/model"
+	"github.com/juju/juju/core/modelmigration"
+	corepermission "github.com/juju/juju/core/permission"
+	"github.com/juju/juju/core/user"
+	accesserrors "github.com/juju/juju/domain/access/errors"
+	"github.com/juju/juju/domain/access/service"
+	"github.com/juju/juju/domain/access/state"
+)
+
+// Coordinator is the interface that is used to add operations to a migration.
+type Coordinator interface {
+	// Add adds the given operation to the migration.
+	Add(modelmigration.Operation)
+}
+
+// RegisterImport registers the import operations with the given coordinator.
+func RegisterImport(coordinator Coordinator, logger logger.Logger) {
+	coordinator.Add(&importOperation{
+		logger: logger,
+	})
+}
+
+// ImportService provides a subset of the access domain
+// service methods needed for model permissions import.
+type ImportService interface {
+	// CreatePermission gives the user access per the provided spec.
+	// If the user provided does not exist or is marked removed,
+	// accesserrors.PermissionNotFound is returned.
+	// If the user provided exists but is marked disabled,
+	// accesserrors.UserAuthenticationDisabled is returned.
+	// If a permission for the user and target key already exists,
+	// accesserrors.PermissionAlreadyExists is returned.
+	CreatePermission(ctx context.Context, spec corepermission.UserAccessSpec) (corepermission.UserAccess, error)
+	// SetLastModelLogin will set the last login time for the user to the given
+	// value. The following error types are possible from this function: -
+	// accesserrors.UserNameNotValid: When the username supplied is not valid. -
+	// accesserrors.UserNotFound: When the user cannot be found. -
+	// modelerrors.NotFound: If no model by the given modelUUID exists.
+	SetLastModelLogin(ctx context.Context, name user.Name, modelUUID coremodel.UUID, time time.Time) error
+}
+
+type importOperation struct {
+	modelmigration.BaseOperation
+
+	logger  logger.Logger
+	service ImportService
+}
+
+// Name returns the name of this operation.
+func (i *importOperation) Name() string {
+	return "import model user permissions"
+}
+
+// Setup implements Operation.
+func (i *importOperation) Setup(scope modelmigration.Scope) error {
+	i.service = service.NewService(
+		state.NewState(scope.ControllerDB(), i.logger))
+	return nil
+}
+
+// Execute the import on the model user permissions contained in the model.
+func (i *importOperation) Execute(ctx context.Context, model description.Model) error {
+	modelUUID := model.Tag().Id()
+	for _, u := range model.Users() {
+		name := user.NameFromTag(u.Name())
+		access := corepermission.Access(u.Access())
+		if err := access.Validate(); err != nil {
+			return errors.Annotatef(err, "importing access for user %q", name)
```

Comment:
**question(maybe issue):**  The import is done in a loop, but an item of the loop can result in an error. 

It means that we can have an incomplete import. Is it ok to have an incomplete import ? 

- If not, it should be kinda transactionnal (at least verify everything which has no side effect in a first loop)
- If it is ok, maybe we should collect all errors in order to have the most complete import and a collection of error (I don't know if juju/errors support std [errors.Join](https://pkg.go.dev/errors#Join) but it is one of it use cases)

Reply:
If any item has an error then the whole import will be cancelled and the model will not be migrated (as far as I understand it). If we get an error we will throw away all the data we have imported so far anyway.